openedx_tagging/core/tagging/models.py

```diff
@@ -1,39 +1,440 @@
+    def __str__(self):
+        """
+        User-facing string representation of a Tag.
+        """
+        return f"Tag ({self.id}) {self.value}"
+
+    def get_lineage(self) -> Lineage:
+        """
+        Queries and returns the lineage of the current tag as a list of Tag.value strings.
 
-    Name-Value pair
-    -----------------
+        The root Tag.value is first, followed by its child.value, and on down to self.value.
 
-    We use a Field-Value Pair representation, where `name` functions
-    as the constant that defines the field data set, and `value` functions
-    as the variable tag lineage that belong to the set.
+        Performance note: may perform as many as TAXONOMY_MAX_DEPTH select queries.
+        """
+        lineage: Lineage = []
+        tag = self
+        depth = TAXONOMY_MAX_DEPTH
+        while tag and depth > 0:
+            lineage.insert(0, tag.value)
+            tag = tag.parent
+            depth -= 1
+        return lineage
 
-    Value lineage
-    ---------------
 
-    `value` are stored as null character-delimited strings to preserve the tag's lineage.
-    We use the null character to avoid choosing a character that may exist in a tag's name.
-    We don't use the Taxonomy's user-facing delimiter so that this delimiter can be changed
-    without disrupting stored tags.
+class Taxonomy(models.Model):
     """
+    Represents a namespace and rules for a group of tags which can be applied to a particular Open edX object.
+    """
+
+    id = models.BigAutoField(primary_key=True)
+    name = models.CharField(
+        null=False,
+        max_length=255,
+        db_index=True,
+        help_text=_(
+            "User-facing label used when applying tags from this taxonomy to Open edX objects."
+        ),
+    )
+    description = models.TextField(
+        null=True,
+        blank=True,
+        help_text=_(
+            "Provides extra information for the user when applying tags from this taxonomy to an object."
+        ),
+    )
+    enabled = models.BooleanField(
+        default=True,
+        help_text=_("Only enabled taxonomies will be shown to authors."),
+    )
+    required = models.BooleanField(
+        default=False,
+        help_text=_(
+            "Indicates that one or more tags from this taxonomy must be added to an object."
+        ),
+    )
+    allow_multiple = models.BooleanField(
+        default=False,
+        help_text=_(
+            "Indicates that multiple tags from this taxonomy may be added to an object."
+        ),
+    )
+    allow_free_text = models.BooleanField(
+        default=False,
+        help_text=_(
+            "Indicates that tags in this taxonomy need not be predefined; authors may enter their own tag values."
+        ),
+    )
+    _object_tag_class = models.CharField(
+        max_length=255,
+        null=True,
+        blank=True,
+        help_text=_(
+            "Python path.module.class of the object tag class associated with this taxonomy.",
```

Comment:
Can you remind me, what is the use case for being able to override `ObjectTag` per taxonomy?

Reply:
I generally like the YAGNI philosophy so I'd say add it when needed :)

Reply:
Cool.. I'll put this code into the discovery in case it's needed later, but will remove it from here.